orttraining/orttraining/python/training/ortmodule/ortmodule.py

```diff
@@ -165,35 +156,40 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                                 missing_keys, unexpected_keys, error_msgs):
         """Override original method to delegate execution to the original PyTorch user module"""
 
-        # PyTorch load_state_dict implementation does not recursively call load_state_dict on its sub-modules. 
-        # Instead, it creates a recursive function and invokes _load_from_state_dict on all child modules.
-        # For the scenario where an ORTModule is a sub-module of another module, loading of the state
-        # dictionary requires the _load_from_state_dict to be overridden to prevent an error.
-        self._module_metadata.original_module._load_from_state_dict(state_dict, prefix, local_metadata, strict,
-                                missing_keys, unexpected_keys, error_msgs)
+        self._module_manager._load_from_state_dict(state_dict, prefix, local_metadata, strict,
+                                                   missing_keys, unexpected_keys, error_msgs)
 
     def named_children(self) -> Iterator[Tuple[str, 'Module']]:
         """Override original method to delegate execution to the original PyTorch user module"""
 
-        yield from self._module_metadata.original_module.named_children()
+        yield from self._module_manager.named_children()
 
     def modules(self) -> Iterator['Module']:
         """Override original method to delegate execution to the original PyTorch user module"""
 
-        yield from self._module_metadata.original_module.modules()
+        yield from self._module_manager.modules()
 
     def named_modules(self, *args, **kwargs):
         """Override original method to delegate execution to the original PyTorch user module"""
 
-        # PyTorch >1.8.1 has an extra arg remove_duplicate that is not present in 1.8.1
-        # To support both, use args and kwargs (since user can call the method with only positional args or kwargs)
-        yield from self._module_metadata.original_module.named_modules(*args, **kwargs)
+        yield from self._module_manager.named_modules(*args, **kwargs)
 
     def add_module(self, name: str, module: Optional['Module']) -> None:
         """Raises a NotImplementedError exception since ORTModule does not support adding modules to it"""
 
         raise NotImplementedError("ORTModule does not support adding modules to it.")
 
+    def _validate_module_type(self, module):
+        """Raises a TypeError if the module is not a torch.nn.Module"""
+
+        if not isinstance(module, torch.nn.Module):
+            raise TypeError(f"ORTModule only support torch.nn.Module as input. {type(module)} is not supported.")
+
+    def _create_module_manager(self, module):
```

Comment:
Refactoring this to a factory class, say `TorchModuleFactory`, minimizes changes to `ORTModule` *implementation*. The user will only modify the factory instance on ORTModule with their own factory.

Reply:
Let me try doing this. Monkey paching a free function like ```TorchModuleFactory``` might be more involved than overriding a method of ```ORTModule```. Ideally, this should be a separate method by itself.

Reply:
My suggestion is making `TorchModuleFactory` as a class, as the design pattern dictates. ONNX implementation will have to extend this factory and replace its instance on `ORTModule` object they instantiate. It will keep being a classic factory though

Reply:
Addressed.